### PR TITLE
Faststream poc

### PR DESCRIPTION
**Pull request by TrangPham**

### CODE REVIEW

**Review by Kilo59 — APPROVED** [review]

File: great_expectations_cloud/agent/agent.py [hunk 1]
```diff
@@ -102,6 +111,82 @@ class Config:
...
+    print(f"Received: {msg}")
+    print(f"GX Agent: {gx_agent}")
+
+    event = EventHandler.parse_event_from_dict(msg)
+    event_context = EventContext(
+        event=event,
+        correlation_id=correlation_id,  # msg.correlation_id,
+        processed_successfully=mock.Mock(),
+        processed_with_failures=mock.Mock(),
+        redeliver_message=mock.Mock(),
```
Comment (allensallinger) [inline great_expectations_cloud/agent/agent.py:126]: Why are the values here set to `mock.Mock()`?

**Review by allensallinger — APPROVED** [review]

File: .python-version [hunk 2]
```diff
@@ -1 +1 @@
-3.10.14
+3.11.9
```
Comment (rreinoldsc) [inline .python-version:1]: Is it best to bring this change in separately?

File: great_expectations_cloud/agent/agent.py [hunk 3]
```diff
@@ -102,6 +111,82 @@ class Config:
         json_loads = orjson_loads
 
 
+# Use Retries = 10, --workers 1
+# Data Context as a Dependency
```
Comment (rreinoldsc) [inline great_expectations_cloud/agent/agent.py:115]: I'm not sure what these comments mean, should they stay?

```suggestion
```

File: great_expectations_cloud/agent/agent.py [hunk 4]
```diff
@@ -10,9 +10,16 @@
 from functools import partial
 from importlib.metadata import version as metadata_version
 from typing import TYPE_CHECKING, Any, Callable, Dict, Final, Literal
+from unittest import mock
```
Comment (rreinoldsc) [inline great_expectations_cloud/agent/agent.py:13]: Shouldn't have any `unittest` imports in production

File: great_expectations_cloud/agent/agent.py (hunk 1 above)
Comment (rreinoldsc) [reply great_expectations_cloud/agent/agent.py:126]: +1

**Review by rreinoldsc — CHANGES_REQUESTED** [review]
`unittest` dep should not be in application code

File: great_expectations_cloud/agent/agent.py [hunk 5]
```diff
@@ -102,6 +111,82 @@ class Config:
...
+        correlation_id=correlation_id,  # msg.correlation_id,
+        processed_successfully=mock.Mock(),
+        processed_with_failures=mock.Mock(),
+        redeliver_message=mock.Mock(),
+    )
+    organization_id = None
+    try:
+        organization_id = gx_agent.get_organization_id(event_context)
+        result = gx_agent._handle_event(event_context)
+    except Exception as e:
```
Comment (rreinoldsc) [inline great_expectations_cloud/agent/agent.py:132]: Can this be a more narrow catch?

File: great_expectations_cloud/agent/agent.py [hunk 6]
```diff
@@ -327,9 +426,9 @@ def _handle_event_as_thread_exit(
                     extra={
                         "event_type": event_context.event.type,
                         "correlation_id": event_context.correlation_id,
-                        "job_duration": result.job_duration.total_seconds()
-                        if result.job_duration
-                        else None,
+                        "job_duration": (
+                            result.job_duration.total_seconds() if result.job_duration else None
```
Comment (rreinoldsc) [inline great_expectations_cloud/agent/agent.py:430]: This seems to be out of scope of your changes, but a `job_duration` of 0 seconds would result in `None`

```suggestion
                            result.job_duration.total_seconds() if result.job_duration else None
```

File: great_expectations_cloud/agent/agent.py [hunk 7]
```diff
@@ -144,43 +229,57 @@ def run(self) -> None:
...
-            print("The GX Agent is ready.")
-            # Open a connection until encountering a shutdown event
-            subscriber.consume(
-                queue=self._config.queue,
-                on_message=self._handle_event_as_thread_enter,
+            broker = RabbitBroker(
+                url=str(self._config.connection_string),
             )
+            app = FastStream(broker)
+            queue = RabbitQueue(name=self._config.queue, durable=True, passive=True)
```
Comment (anthonyburdi) [inline great_expectations_cloud/agent/agent.py:254]: Is passive to make this idempotent? Seeing "A passive declare simply checks that the entity with the provided name exists. If it does, the operation is a no-op." from: https://www.rabbitmq.com/client-libraries/java-api-guide#passive-declaration

File: great_expectations_cloud/agent/agent.py [hunk 8]
```diff
@@ -144,43 +229,57 @@ def run(self) -> None:
...
+            print("Queue is valid.")
+
+            # FastStream declares default exchange if not provided
+            @broker.subscriber(queue, retry=MAX_DELIVERY)
+            async def handle_me(
```
Comment (anthonyburdi) [inline great_expectations_cloud/agent/agent.py:259]: Is this used for checking to make sure rabbitmq is set up and ready to handle messages?

File: great_expectations_cloud/agent/agent.py [hunk 9]
```diff
@@ -144,43 +229,57 @@ def run(self) -> None:
...
             print("Received request to shut down.")
-        except (SubscriberError, ClientError):
+        except (
+            SubscriberError,
+            ClientError,
+            asyncio.exceptions.CancelledError,
+            aiormq.exceptions.ChannelAccessRefused,
+        ):
             print("The connection to GX Cloud has encountered an error.")
+            print(traceback.format_exc())
```
Comment (anthonyburdi) [inline great_expectations_cloud/agent/agent.py:277]: Is there any reason to not surface this error in logs? I can't think of one, just checking.

**Review by anthonyburdi — COMMENTED** [review]
Just a few questions in addition to those already posed.

File: great_expectations_cloud/agent/agent.py (hunk 6 above)
Comment (rreinoldsc) [reply great_expectations_cloud/agent/agent.py:428]: out of scope

File: great_expectations_cloud/agent/agent.py (hunk 9 above)
Comment (rreinoldsc) [reply great_expectations_cloud/agent/agent.py:275]: Pushing fix

Comment (rreinoldsc) [reply great_expectations_cloud/agent/agent.py]: addressed in 5fcd42b

File: great_expectations_cloud/agent/agent.py (hunk 4 above)
Comment (rreinoldsc) [reply great_expectations_cloud/agent/agent.py]: removed in 88e2de4

**Review by rreinoldsc — APPROVED** [review]

**Pull request closed without merging**